### PR TITLE
Distcp ng handle directory structure copy 

### PR DESCRIPTION
**Pull request by tuGithub**
also set the owner of the hive table as the flow executor

### CODE REVIEW

**Comment by tuGithub** [comment]
@ibuenros , @chavdar , please review it when you have time

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 1]
```diff
@@ -95,6 +96,25 @@
    */
   public static final String COPY_TARGET_TABLE_ROOT =
       HiveDatasetFinder.HIVE_DATASET_PREFIX + ".copy.target.table.root";
+  
+  /**
+   * These two options, in pair, specify the output location of the data files on copy
+   * {@link #COPY_TARGET_TABLE_PREFIX_TOBE_REPLACED} specified the prefix of the path (without Scheme and Authority ) to be replaced
+   * {@link #COPY_TARGET_TABLE_PREFIX_REPLACEMENT} specified the replacement of {@link #COPY_TARGET_TABLE_PREFIX_TOBE_REPLACED}
+   * <p>
+   * for example, if the data file is hdfs://ltx1-unonn01.grid.linkedin.com:9000/data/databases/ANET/ANET_CONTACTS_MANIFEST/1462826996005-PT-2085190/part-00000.avro ,
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Please remove this from the comment (directory structure is internal information), and squash the commits to make it disappear from git history.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 2]
```diff
@@ -95,6 +96,25 @@
...
+   * for example, if the data file is hdfs://ltx1-unonn01.grid.linkedin.com:9000/data/databases/ANET/ANET_CONTACTS_MANIFEST/1462826996005-PT-2085190/part-00000.avro , 
+   * {@link #COPY_TARGET_TABLE_PREFIX_TOBE_REPLACED} is /data/databases
+   * {@link #COPY_TARGET_TABLE_PREFIX_REPLACEMENT} is /data/databases/_parallel
+   * 
+   * then, the output location for that file will be
+   * $targetFs/data/databases/_parallel/ANET/ANET_CONTACTS_MANIFEST/1462826996005-PT-2085190/part-00000.avro
+   * </p>
+   */
+  public static final String COPY_TARGET_TABLE_PREFIX_TOBE_REPLACED =
+	      HiveDatasetFinder.HIVE_DATASET_PREFIX + ".copy.target.table.prefix.tobe.replaced";
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Can you use fewer dots? Maybe `.copy.target.table.prefixToBeReplaced`?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 3]
```diff
@@ -421,6 +450,10 @@ private Table getTargetTable(Table originTable, Path targetLocation) throws IOEx
 
       targetTable.setDbName(this.targetDatabase);
       targetTable.setDataLocation(targetLocation);
+      /*
+       * Need to set the table owner as the flow executor
+       */
+      targetTable.setOwner(UserGroupInformation.getCurrentUser().getShortUserName());
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: You should also do this for partitions.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 4]
```diff
@@ -800,17 +832,20 @@ public Path getTargetPath(Path sourcePath, FileSystem targetFs, Optional<Partiti
...
-      Preconditions.checkArgument(PathUtils.isAncestor(this.dataset.tableRootPath.get(), sourcePath),
-          "When moving paths to a new root, all locations must be descendants of the table root location. "
-              + "Table root location: %s, file location: %s.", this.dataset.tableRootPath, sourcePath);
-
-      Path relativePath = PathUtils.relativizePath(sourcePath, this.dataset.tableRootPath.get());
-      return targetFs.makeQualified(new Path(this.targetTableRoot.get(), relativePath));
-    } else {
+    
+    // both prefixs must be present as the same time
+    if (this.targetTablePrefixTobeReplaced.isPresent() || this.targetTablePrefixReplacement.isPresent()){
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Is this backwards compatible with `targetTableRoot`?

Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Can you write unit tests for this functionality? Thanks.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 1 above)
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: done

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 2 above)
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: done

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 3 above)
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Partition can not set owner

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 4 above)
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Done for unit testing.

Added back the logic to handle targetTableRoot

**Comment by tuGithub** [comment]
@ibuenros , updated the PR with comments

**Comment by ibuenros** [comment]
LGTM

**Pull request merged**